### ISSUE TITLE
End columns (and eventually fix suggestions)

### ISSUE THREAD

**Issue by ngzhian** [koalaman/shellcheck#1262]
Originally from https://github.com/koalaman/shellcheck/pull/1239#issuecomment-397152045 (filing this as a separate issue to more easily track what to do to move the end column effort forward, hope you don't mind!)

- [x] Set end positions everywhere
- [x] Delete the zero-width functions so they won't be used in the future
- [ ] Go over all warnings to ensure they have meaningful end positions
- [ ] Expose end positions in the tty and gcc output modes, and on shellcheck.net
- [ ] Inevitable extras that come up, like how T_WhileExpression throws away information about the while keyword, so no test is able to highlight just the while

**Comment by koalaman** [comment]
The first two are mostly done, and the tty output now shows ranges if they're reasonably short and on the same line.

**Comment by ngzhian** [comment]
Q: how do we want to expose fixes?
Maybe adding a fix field to PositionedComment, and we can build this up where the warning is found. An this can be exposed in output formats that make sense to show this, e.g. JSON.
Alternatively we can have a module that will look at PositionedComment of a file and build another data type to represent fixes.

**Comment by koalaman** [comment]
I definitely like the idea of building the fix where it's found. I think it's reasonable to have a SuggestedFix and add it to the PositionedComment (or `Maybe SuggestedFix`).

Immediately before, it would probably also be a good idea to make everything in ShellCheck.Interface more opaque, so that we can add fields to PositionedComment without breaking [hadolint](https://github.com/hadolint/hadolint) again. I added #1300 for this.

[error-prone](https://github.com/google/error-prone) is the only auto-fixing tool I'm vaguely familiar with. Their idea of a fix was a list of (StartPos, EndPos, ReplacementString) tuples per suggestion, e.g. `fgrep foo -> [ (0,1, ""), (6,6, "-F ") ]`, and afaik it was up to the consumer to unify multiple fixes affecting the same line. That seems like a pretty reasonable approach, though it's worth looking into how other tools do it.

One fancy possibility once this is done is a `diff` output format to let you do e.g.  `shellcheck -f diff file.sh | patch`

**Comment by matthewpersico** [comment]
FYI, I currently have an emacs function [shellcheck-disable-error-at-point](https://gist.github.com/matthewpersico/eeeb263acd42570ba7abe1912ad5f506) that will, if you press 'd' in the shellcheck window list, will manipulate the related text and add the correct disable statement in the line above the error point. If you guys are thinking of providing the suggested replacement text, that would be awesome; I could then write ```shellcheck-correct-error-at-point```.

**Comment by ngzhian** [comment]
@koalaman what do you think of my previous comment?

**Comment by koalaman** [comment]
I like the two step approach in theory, but I'm worried it'll devolve into a set of narrow ad-hoc fixes like `DeleteFirstCharOfCommandAndAlsoAddFlag "F"` (replacing `fgrep foo` with `grep -F foo`). However, maybe I'm totally wrong and there's a simple set of constructs that apply to most fixes.

I looked at a randomized selection of issues that appear auto-fixable:

```
# 1
DEFS=( -DBADVPN_THREAD_SAFE=0 -DBADVPN_LINUX -DBADVPN_BREACTOR_BADVPN -D_GNU_SOURCE )                             
       ^--------------------^ SC2191: The = here is literal. To assign by index, use ( [index]=value ) with

# 2
    "${CC}" -c ${CFLAGS} "${INCLUDES[@]}" "${DEFS[@]}" "${SRCDIR}/${f}" -o "${obj}"
               ^-------^ SC2086: Double quote to prevent globbing and word splitting.

# 3
cd "$DIR"
^-------^ SC2164: Use 'cd ... || exit' or 'cd ... || return' in case cd fails.

# 4
CUR_HASH=`$pm2 prettylist | grep "revision" | cut -d: -f2 | tr -d " ,'"`
         ^-- SC2006: Use $(...) notation instead of legacy backticked `...`.

# 5
cd ..
^---^ SC2103: Use a ( subshell ) to avoid having to cd back.

# 6
GIT_VERSION_MAJOR=$(printf $GIT_VERSION | cut -d. -f1)
                           ^----------^ SC2059: Don't use variables in the printf format string. Use printf "..%s.." "$foo".

# 7
output:$(xrandr --verbose | awk '$1 == "Brightness:" {print $2; exit}')
       ^-- SC2046: Quote this to prevent word splitting.


# 8
            NIXCFG_HOST=$(echo $(find -type d -name "${NIXCFG_HOST}")/configuration.sh)
                        ^-- SC2116: Useless echo? Instead of 'cmd $(echo foo)', just use 'cmd foo'.
                                 ^--^ SC2185: Some finds don't have a default path. Specify '.' explicitly.

#9
which git > /dev/null 2>&1 || __bootstrap_git
^---^ SC2230: which is non-standard. Use builtin 'command -v' instead.

#10
    local source="${BUILT_PRODUCTS_DIR}/$(basename "$1")"
          ^----^ SC2155: Declare and assign separately to avoid masking return values.
```

Here are the reusable type fixes that apply:

1. `SurroundWithQuotes` would work ok, but in the case of `a=$b` it should at least take a start token and end token to surround, and not just a single token.

2. `SurroundWithQuotes`

3. `AppendAfterToken x " || exit"`

4. Not sure

5. `InsertBeforeToken x "(", InsertAfterToken y ")"`

6. `ReplaceToken x "...%s...", AppendAfterToken y " $GIT_VERSION"`

7. `SurroundWithQuotes`

8. Not sure

9. `ReplaceToken x "command -v"`

10. Not sure

There's also totally the possibility of just adding source level access to let each check generate a fix, or alternatively to have them return a "FixGenerator" function that optionally runs with access to such data if the fix is needed.

What's your opinion on 4, 8 and 10? What would the check emit in the two-step approach?

**Comment by ngzhian** [comment]
Making up an imaginary API:

```
-- low level
replace_start :: Token -> String -> String -> Fix -- replace matching string at the start of token
replace_end :: Token -> String -> String -> Fix -- replace matching string at the end of token
-- OR
replace_index :: Token -> Int -> Int -> String -> Fix -- replace a range within Token by a String

-- helpers
insert_before :: Token -> String -> Fix -- implemented in terms of replace_*
insert_after :: Token -> String -> Fix -- implemented in terms of replace_*
surround :: Token -> String -> Fix -- implemented in terms of insert_before and insert_after
surround2 :: Token -> Token -> String -> Fix -- for surrounding multiple tokens, implemented with insert_before and insert_after
```

4, can be
- ``replace_start token "`" "$("`` and ``replace_end token "`" ")"``, or
- `replace_index token 0 1 "$("` and `replace_index token (end_column-1) end_column ")"`


8, can be `insert_after find_token " . "`, and maybe replace start and end with empty strings to remove the `$(echo ... )`

10, is tricky, we can maybe do an `insert_after source_token "\nsource"`, and in this case we can get the value "source" at warning time because this will be a simple variable name.

> There's also totally the possibility of just adding source level access to let each check generate a fix, or alternatively to have them return a "FixGenerator" function that optionally runs with access to such data if the fix is needed.

This is doable too, instead of generating a new data type `Fix`, we can use some helper methods to create lambdas that can be passed to the formatter.

I guess the most tricky part I haven't thought much about is applying multiple fix that changes the position of the token. Whichever way we go with that makes this easy to do is probably the better way.

**Comment by matthewpersico** [comment]
How do you envision all the actual fixing integrating with editors?

**Comment by ngzhian** [comment]
I think to begin with it could be something like a diff/patch replacement on lines
e.g. in the json output it could have a replacementLine :: [String], one replacement for each line that the warning is for.

**Comment by matthewpersico** [comment]
The tricky part is what happens if there are multiple edits per line. You almost ha e to have one entry with all of them or else as you apply them, subsequent ones may not match anymore.

**Comment by ngzhian** [comment]
Yea good point, i was thinking of how the tool that applies the replacement can keep track of any adjustments needed. But that kind of complicates matters esp for editor integration. The easy way out is to only apply 1 fix at a time, and then run shellcheck again on the fixed output, until we get to the end.

**Comment by matthewpersico** [comment]
Which kinda what happens now in an editor as I correct issues one by one

**Comment by ngzhian** [comment]
I think in the editor you should look at each fix and make sure it's fine.
When done in an automated way we can then have shellcheck output a patch
file. Or even just completely overwrite the file?

**Comment by koalaman** [comment]
I haven't had a chance to evaluate the API yet, but the ability to hit "Fix" in an interactive editor is really the least interesting use case, because it's only marginally more work to fix it manually.

I imagine the more useful case will be "Fix All" to auto-fix all instances in the file/dir, and for code review tool integration where you can hit "Suggest fix" to automatically submit a comment complete with a code replacement suggestion.

Merging multiple fixes per line will indeed be annoying when they affect the same ranges, but merge conflicts always are. Using character range replacements rather than line replacements helps with this.

**Comment by matthewpersico** [comment]
Hmm - "Fix All" sounds like a good idea, but I can tell you from experience there is usually once or twice in a file where I end up not taking the most common advice - quoting a variable. I would sit there and bounce on two keys: next error in the shellcheck message buffer and 'F' to run the emacs function to fix, or 'D' to stick in the disable directive.

**Comment by koalaman** [comment]
b1aeee56 adds a `diff` output which does allow you to "fix all" by piping to `patch` or `git apply`. Any editor functionality for interactively applying would be up to individual editor plugins.
